### PR TITLE
Ope collections sources

### PR DESCRIPTION
**Pull request by osahon-okungbowa**
TASKS TACKLED
- fixed metadata request headers retrieval
- resource headers now available in datajson transformer output
- fixed backtracking issue caused by 'false' referer head AKA bug where collections and sources are the same
- implemented simple throttling for requests of resources which are outside of scrapy's control. this solves rate limit errors
- other performance and resilience improvements

### CODE REVIEW

**Comment by osahon-okungbowa** [comment]
@nightsh 2 collections, 1 source in ocr now. Pls kindly merge by tomorrow morning. so P4 (OELA)can benefit from all improvements. Thank you :+1:

File: edscrapers/scrapers/base/helpers.py [hunk 1]
```diff
@@ -4,8 +4,11 @@
 import datetime
 import hashlib
 import urllib.parse
+import itertools
 import requests
 import bs4
+import ratelimit
```
Comment (nightsh) [inline edscrapers/scrapers/base/helpers.py:10]: I wonder if there's any point extracting this into more generic helpers, so they'd be available for other parsers as well... :thinking:

File: edscrapers/scrapers/base/helpers.py [hunk 2]
```diff
@@ -81,16 +93,22 @@ def get_meta_value(soup, meta_name):
...
         raw_headers = requests.head(url).headers
     else:
         raw_headers = requests.head(urljoin(source_url, url)).headers
 
-    headers['content-type'] = raw_headers['Content-Type']
-    headers['last-modified'] = raw_headers['Last-Modified']
-    headers['content-length'] = raw_headers['Content-Length']
+    headers['content-type'] = raw_headers.get('Content-Type', None)
+    headers['last-modified'] = raw_headers.get('Last-Modified', None)
+    headers['content-length'] = raw_headers.get('Content-Length', None)
```
Comment (nightsh) [inline edscrapers/scrapers/base/helpers.py:111]: same as above, we should probably make this into a helper at some point, soon

File: edscrapers/scrapers/base/helpers.py [hunk 3]
```diff
@@ -235,8 +279,13 @@ def extract_dataset_source_from_url(source_url, namespace):
     
     source = Source()
     source['source_url'] = source_url
-    source['source_title'] = str(soup_parser.head.\
+    # get the Source title
+    if soup_parser.head.find(name='title'):
+        source['source_title'] = str(soup_parser.head.\
                             find(name='title').string).strip()
+    else:
+        source['source_title'] = '[no title]'
```
Comment (nightsh) [inline edscrapers/scrapers/base/helpers.py:287]: I know it's a precaution, but have you ever encountered such a thing?

File: edscrapers/scrapers/ocr/parsers/ocr_state_national_estimates_parser2.py [hunk 4]
```diff
@@ -29,8 +29,8 @@ def parse(res) -> dict:
         collection = h.extract_dataset_collection_from_url(collection_url=res.url,
                                         namespace="all",
                                         source_url=\
-                                        str(res.request.headers[str(b'Referer',
-                                                                    encoding='utf-8')], 
+                                        str(res.request.headers.get(str(b'Referer',
```
Comment (nightsh) [inline edscrapers/scrapers/ocr/parsers/ocr_state_national_estimates_parser2.py:32]: we should get into the habit of never accessing dict keys directly...

**Pull request merged**